openlr/openlr_decoder.cpp

```diff
@@ -191,4 +310,107 @@ void OpenLRDecoder::Decode(vector<LinearSegment> const & segments, uint32_t cons
+  {
+    if (!DecodeOne(segments[i], m_indexes.back(), graph, infoGetter, paths[i], stat))
+      ++stat.m_routesFailed;
+    ++stat.m_routesHandled;
+
+    if (i % 100 == 0 || i == segments.size() - 1)
+    {
+      LOG(LINFO, (i, "segments are processed in",
+                  timer.ElapsedSeconds(),
+                  "seconds"));
+      timer.Reset();
+    }
+  }
+
+  LOG(LINFO, ("Total routes handled:", stat.m_routesHandled));
+  LOG(LINFO, ("Failed:", stat.m_routesFailed));
+  LOG(LINFO, ("No candidate lines:", stat.m_noCandidateFound));
+  LOG(LINFO, ("Wrong dnp:", stat.m_dnpIsZero));
+  LOG(LINFO, ("Wrong offsets:", stat.m_wrongOffsets));
+  LOG(LINFO, ("No shortest path:", stat.m_noShortestPathFound));
+}
+
+bool OpenLRDecoder::DecodeOne(LinearSegment const & segment, Index const & index, Graph & graph,
+                              RoadInfoGetter & infoGetter, DecodedPath & path, v2::Stats & stat)
+{
+  // TODO(mgsergio): Scale indexes.
+
+  graph.ResetFakes();
+
+  double const kPathLengthTolerance = 0.30;
+  uint32_t const kMaxJunctionCandidates = 10;
+  uint32_t const kMaxProjectionCandidates = 5;
+
+  path.m_segmentId.Set(segment.m_segmentId);
+
+  auto const & points = segment.GetLRPs();
+  vector<vector<Graph::EdgeVector>> lineCandidates;
+  lineCandidates.reserve(points.size());
+  LOG(LDEBUG, ("Decoding segment:", segment.m_segmentId, "with", points.size(), "points"));
+
+  CandidatePointsGetter pointsGetter(kMaxJunctionCandidates, kMaxProjectionCandidates, index, graph);
+  CandidatePathsGetter pathsGetter(pointsGetter, graph, infoGetter, stat);
+
+  if (!pathsGetter.GetLineCandidatesForPoints(points, lineCandidates))
+    return false;
+
+  vector<Graph::EdgeVector> resultPath;
+  PathsConnector connector(kPathLengthTolerance, graph, stat);
+  if (!connector.ConnectCandidates(points, lineCandidates, resultPath))
+    return false;
+
+  // TODO(mgsergio): Make a concant iterator
+  Graph::EdgeVector route;
+  for (auto const & part : resultPath)
+    route.insert(end(route), begin(part), end(part));
+
+  uint32_t requiredRouteDistanceM = 0;
+  for (auto const & p : points)
+    requiredRouteDistanceM += p.m_distanceToNextPoint;
+  uint32_t actualRouteDistanceM = 0;
+  for (auto const & e : route)
+    actualRouteDistanceM += EdgeLength(e);
+
+  auto const scale = static_cast<double>(actualRouteDistanceM) / requiredRouteDistanceM;
+  LOG(LDEBUG, ("actualRouteDistance:", actualRouteDistanceM,
+               "requiredRouteDistance:", requiredRouteDistanceM, "scale:", scale));
+
+  auto const positiveOffset =
+      static_cast<uint32_t>(segment.m_locationReference.m_positiveOffsetMeters * scale);
+  auto const negativeOffset =
+      static_cast<uint32_t>(segment.m_locationReference.m_negativeOffsetMeters * scale);
+
+  if (positiveOffset + negativeOffset >= requiredRouteDistanceM)
+  {
+    ++stat.m_wrongOffsets;
+    LOG(LINFO, ("Wrong offsets for segment:", segment.m_segmentId));
```

Comment:
optional: consider to just drop off offsets, i.e. if positiveOffset + negativeOffset >= requiredDistance, try to build route without offsets at all, and take a look at the quality metric.

Reply:
Not this time.